certbot/auth_handler.py

```diff
@@ -47,242 +43,137 @@ def __init__(self, auth, acme_client, account, pref_challs):
+        if not achalls:
+            return authzrs
+
+        # Starting to now, challenges will be cleaned at the end no matter what.
+        with error_handler.ExitHandler(self._cleanup_challenges, achalls):
+            # To begin, let's ask the authenticator plugin to perform all challenges.
+            try:
+                resps = self.auth.perform(achalls)
+
+                # If debug is on, wait for user call before starting the verification process.
+                logger.info('Waiting for verification...')
+                config = zope.component.getUtility(interfaces.IConfig)
                 if config.debug_challenges:
+                    notify = zope.component.getUtility(interfaces.IDisplay).notification
                     notify('Challenges loaded. Press continue to submit to CA. '
                            'Pass "-v" for more info about challenges.', pause=True)
+            except errors.AuthorizationError as error:
+                logger.critical('Failure in setting up challenges.')
+                logger.info('Attempting to clean up outstanding challenges...')
+                raise error
+            # All challenges should have been processed by the authenticator.
+            assert len(resps) == len(achalls), 'Some challenges have not been performed.'
 
-                # Send all Responses - this modifies achalls
-                self._respond(aauthzrs, resp, best_effort)
+            # Inform the ACME CA server that challenges are available for validation.
+            for achall, resp in zip(achalls, resps):
+                self.acme.answer_challenge(achall.challb, resp)
 
-        # Just make sure all decisions are complete.
-        self.verify_authzr_complete(aauthzrs)
+            # Wait for authorizations to be decided.
+            authzrs_still_pending = self._poll_authorizations(authzrs, max_retries, best_effort)
 
-        # Only return valid authorizations
-        ret_val = [aauthzr.authzr for aauthzr in aauthzrs
-                   if aauthzr.authzr.body.status == messages.STATUS_VALID]
+            # All authorizations should be decided now. If not, we reached the max retries.
+            if authzrs_still_pending:
+                raise errors.AuthorizationError('All challenges could not be validated on time.')
 
-        if not ret_val:
-            raise errors.AuthorizationError(
-                "Challenges failed for all domains")
+            # Keep validated authorizations only. If there is none, no certificate can be issued.
+            authzrs_validated = [authzr for authzr in authzrs
+                                 if authzr.body.status == messages.STATUS_VALID]
+            if not authzrs_validated:
+                raise errors.AuthorizationError('All challenges have failed.')
 
-        return ret_val
+            return authzrs_validated
 
-    def _choose_challenges(self, aauthzrs):
+    def _poll_authorizations(self, authzrs, max_retries, best_effort):
```

Comment:
It works, but the return value of this function seems a little awkward to me. Do you think we can use something else instead?

Reply:
I replaced by an exception handling, the method does not return something anymore.